Ladder step4 3

junibibi opened the pull request:
-플레이어 이름이 출력되게끔 함수 추가및 수정
-오류처리 코드추가
-객체역할 분담

JurdyPark committed:
함수추가 및 오류체크케이스 파일생성

사용자 이름 입력받는 함수 추가
사다리층 입력받는 함수 이름수정
오류체크케이스 파일 생성

JurdyPark committed:
메세지출력부분 파일 생성하고 사다리넓이 수정

메세지출력하는 부분을 파일로 분리
사다리 넓이를 최대입력글자수인 5개로 수정

JurdyPark committed:
오류를 확인하는 함수 생성

플레이어와 사다리층 입력오류를 체크해서
오류가있으면 오류메세지를보여주고 프로그램종료
오류가없으면 값을 리턴하도록 만들었습니다

오류파일의 이름수정

JurdyPark committed:
플레이어이름 출력하는 함수 생성

사다리하단부분에 플레이어이름 출력하는 함수 만듬.
이부분에 오류가체크된 플레이어이름이 필요해서
이름값만 필요하기 때문에 전역번수를 만들어 리턴하기 전에 그 값을 저장했다가 사용했습니다.

JurdyPark committed:
이름속성을 갖는 참여자 구조체 생성

플레이어 이름 속성을 갖는 구조체를 만들고
메인 코드에 적용시킴

JurdyPark committed:
사다리 발판 하나를 표현하는 LadderStep struct를 작성.

사다리 발판 하나를 표현하는 구조체를 생성하고
함수명을 변경했습니다.

JurdyPark committed:
사다리 한층 구조체 생성

사다리한층에 필요한 데이터를 만드는 함수를 ladderFollr로 분리시켰습니다.

JurdyPark committed:
Ladder 구조체 생성

사다리 가로데이터들을 만드는 함수를 분리했습니다.

JurdyPark committed:
ladderStep구조체에 함수 추가

사다리발판 모양을 만드는 함수를 사다리발판 구조체에 추가했습니다.

ladderStep의 메소드가 추가되면서 구조체의 내부에 생성했던 인스턴스를
다른 구조체에서도 쓰이게 되서 main파일 상단으로 위치를 변경하고 이름을 수정했습니다.

JurdyPark committed:
ladderFloor구조체에 함수 추가

사다리한층의 모양을 만드는 함수를 사다리층 구조체에 추가했습니다.

ladderFlooe에 메소드가 추가되면서 구조체의 내부에 생성했던 인스턴스를
다른 구조체에서도 쓰이게 되서 main파일 상단으로 위치를 변경하고 이름을 수정했습니다.

JurdyPark committed:
ladder구조체에 함수 추가

사다리 가로모양을 만드는 함수를 사다리 구조체에 추가했습니다.

ladder의 메소드가 추가되면서 내부에 생성했던 인스턴스를
다른 구조체에서도 쓰이게 되서 main파일 상단으로 위치를 변경하고 이름을 수정했습니다.

JurdyPark committed:
InputView 구조체 생성

사용자가 입력력을 받는 함수 두개를 InputView 구조체를 만들어 분리하였습니다.

JurdyPark committed:
ResultView 구조체생성

사다리를 프린트하는 함수를 ResultView 구조체로 분리했습니다.
함수명과 파라미터명을 수정하였습니다.

JurdyPark committed:
InputView구조체에 함수추가/LadderGame구조체 생성

InputView구조체에 입력된값 오류처리하는 함수추가 하였습니다.
LadderGame구조체에는 사다리모양이 완성되는 메소드와 플레이어 이름출력하는 메소드를 추가하였습니다.

godrm commented on LadderGame/LadderGame/InputView.swift:
```diff
@@ -0,0 +1,71 @@
+
+struct InputView {
+    
+    var playerNames = Array<String>()
```
객체 속성을 만들때는 private 접근 권한을 주는 것을 우선적으로 고려하세요.

godrm replied:
여전히 private 권한이 지정되지 않은 것 같은데요?

godrm commented on LadderGame/LadderGame/InputView.swift:
```diff
@@ -0,0 +1,71 @@
+        } catch UserInputError.incorrectPlayerData {
+            exit(0)
+        } catch {
+            print("모든상황에러")
```
InputView 역할과 책임을 어디까지 하는게 좋을까요?
main() 부분을 제외하고 exit()로 종료하거나 print()로 에러 문구를 출력하는 것은 지양하세요.

godrm commented on LadderGame/LadderGame/Ladder.swift:
```diff
@@ -0,0 +1,37 @@
+            horizontalLadders[index] = ladderFloor.makeFloorLadders(data: makeData[index], ladders: horizontalLadders[index])
+        }
+        return horizontalLadders
+    }
```
Ladder가 사다리 전체를 표현하는 거라면 속성을 갖고 있는것과 지금처럼 만들어서 return 하는 것과 어떤 차이점이 있을까요?
사다리 데이터를 갖고(저장하고) 있어야 할 필요성은 없나요?

junibibi replied:
저장프로퍼티를 만들면 필요할때마다 저장된 값을 꺼내서 쓸수있고
함수에서 바로 return 하면 값이 필요할때마다 함수를 호출해줘야하는 차이가 있는거 같습니다.
데이터를 저장하고 있는게 효율적인거 같습니다.

godrm commented on LadderGame/LadderGame/Ladder.swift:
```diff
@@ -0,0 +1,37 @@
+        return horizontalData
+    }
+    
+    func makeHorizontalLadders(makeData: [[Bool]]) -> [[String]]{
```
`[[Bool]]`를 출력을 위해서 `[[String]]`으로 바꾸는 것은 사다리 객체가 하지 않아도 될 것 같습니다. 
출력을 위해서 하는 것은 출력 객체 혹은 출력 형식을 담당하는 객체가 있는게 좋습니다. 
그리고 출력하는 방식도 `-` 가 아니라 `-----` 로 바뀌지 않았나요?

godrm commented on LadderGame/LadderGame/LadderPlayer.swift:
```diff
@@ -0,0 +1,14 @@
+
+
+struct LadderPlayer {
+    var names = Array<String>()
```
LadderPlayer는 이름을 여러 개 가질 수 있는건가요?
한 사람 이름만 표현하는게 아닌가요?

junibibi replied:
게임에 참여하는 사람의 이름을 전부 담는걸로 생각 했는데..
한사람 이름만 표현해야 하나요??

godrm replied:
객체 인스턴스 하나가 한 명을 표현하고
여러명이면 배열로 처리하는게 좋지 않을까요?

godrm commented on LadderGame/LadderGame/message.swift:
```diff
@@ -0,0 +1,33 @@
+        print(Type.peopleNum.inputMessage)
+    case .ladderNum:
+        print(Type.ladderNum.inputMessage)
+    }
```
데이터를 표현하는 객체가 직접 출력을 하는 코드를 포함하고 있을 필요는 없습니다. 
어디에 어떻게 출력할지는 별도 객체가 하는게 좋습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,140 +8,21 @@
+    let data = ladder.makeLadderData(peopleNum: inputView.checkPlayerError().count, ladderNum: inputView.checkLadderError())
+    let horizontal = ladder.makeHorizontalLadders(makeData: data)
+    ladderGame.increaseByladderLine(ladders: horizontal)
+    ladderGame.showPlayerName(playerName: inputView.playerNames)
```
전체적인 구조에서 입력 > 처리 > 출력 흐름이 보여야 하는데 그게 명확하지 않네요.
여기에 있는 객체들은 서로 추상화하고 있는 수준이 동일해서 서로 다른 객체에 독립적으로 동작해야 합니다.
앞 단계가 다음 단계로 꼭 필요한 값만 전달할 수 있으면 됩니다.

godrm reviewed (commented):
struct 나 enum 타입을 사용해서 구현한 내용은 긍정적입니다.
구조체에 메소드들이 어떤 단위로 역할을 나눠야 하는지 다음 피드백을 확인해보세요.

JurdyPark committed:
showmessage함수를 InputView구조체로 이동

사다리게임 실행시 문구를보여주는 함수를 InputView구조체로 이동시켜
정의와 출력으로 분리했습니다.

JurdyPark committed:
Ladder구조체 메소드 수정

Ladder구조체안에 저장프로퍼티를 생성하여 메소드가 return하지 않고 값을 저장하게끔 수정.

JurdyPark committed:
struct LadderPlayer  수정

LadderPlayer 프로퍼티 자료형 수정하였습니다.

JurdyPark committed:
OutputForm객체 생성

출력형식을 변경하는 구조체를 만들었습니다.

JurdyPark committed:
main파일 수정

프로그램 종료와 print문을 main파일에서 실행하 수정하였습니다.

junibibi commented:
피드백반영하여 수정하였습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -11,18 +11,32 @@ import Foundation
 let ladderFloor = LadderFloor()
-let ladder = Ladder()
+var ladder = Ladder()
 let ladderGame = LadderGame()
```
여기 작성한 변수도 스코프를 playGame() 내부에서만 참조하도록 옮겨주세요.
이제는 이렇게 스코프가 전역으로 잡히는 변수나 상수를 사용할 일은 거의 없을 겁니다.

godrm commented on LadderGame/LadderGame/LadderGame.swift:
```diff
@@ -0,0 +1,39 @@
+            showName = showName + name
+        }
+        print(showName)
+    }
```
여기 LadderGame 객체는 사다리 게임을 추상화한 역할이기 보다는 출력을 하고 있으니 OutputView 이름이 더 어울릴 것 같습니다.

JurdyPark committed:
outputView 구조체 생성

플레이어 이름을 출력하는 함수를 outputView구조체 생성하여 분리

JurdyPark committed:
필요없는 변수 삭제

사용하지않는 변수 삭제

JurdyPark committed:
구조체 내부에서만 사용되는 메서드에 private 권한 지정

JurdyPark committed:
main파일 수정

인스터스를생성 하지 않고 값을 참조하기위해 인스턴스메소드를 타입메소드로 수정.
main파일-전역에서 생성한 변수/상수를 playGame()함수 내부에서 참조하도록 수정.

junibibi commented:
피드백반영하여 수정했습니다.

godrm commented on LadderGame/LadderGame/OutputView.swift:
```diff
@@ -0,0 +1,27 @@
+            showName = showName + name
+        }
+        print(showName)
+    }
```
여기 OutputView 와 ResultView 가 따로있네요? 같은 역할을 하는 게 아닐까요?

JurdyPark committed:
역할이 중복된 구조체 삭제

OutputView구조체랑 ResultView구조체의 역할이 중복되어
OutputView구조체를 삭제하고 ResultView에 포함시켰습니다.

junibibi commented:
수정했습니다

godrm commented:
이번 스탭에서 해야할 것 중에 남은 것은 ResultView에 접근하는 곳이 main 에서만 접근하도록 하는 겁니다.
입력 > 처리 > 출력 을 담당하는 객체가 모두 동등하다면 접근하는 곳도 모두 동일한 수준이어야 합니다.
이 부분을 다음 단계를 진행하면서 개선해보세요.
